### PR TITLE
Implement class balancing features for YOLO

### PR DESCRIPTION
**Pull request by liam1472**
- Add pos_weight support for BCEWithLogitsLoss in v8DetectionLoss
- Implement WeightedRandomSampler in DataLoader for batch balancing
- Add cls_weights parameter to model.yaml configuration
- Add calculate_class_weights utility function for automatic weight calculation
- Update DetectionTrainer to support class balancing parameters
- Add comprehensive unit tests for all class balancing features
- Add example script demonstrating usage of class balancing features
- Support both manual class weights and automatic calculation
- Maintain backward compatibility with existing configurations

Features implemented:
1. Weighted Loss: pos_weight in BCEWithLogitsLoss penalizes under-represented classes
2. WeightedRandomSampler: Ensures balanced class distribution in training batches
3. Configuration Support: cls_weights parameter in YAML configs
4. Automatic Calculation: Inverse frequency weighting when cls_weights=True

Usage examples:
- model.train(cls_weights=True)  # Auto-calculate weights
- model.train(cls_weights=[1.0, 2.0, 1.5])  # Manual weights
- Use cls_weights in model.yaml configuration files

Requested by: @liam1472
Devin run: https://app.devin.ai/sessions/5255c3ad6e2e4096b28e9948433c0ecd

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds robust class balancing features to Ultralytics YOLO, including automatic and manual class weighting, improved data sampling, and comprehensive tests and examples. 🏋️‍♂️

---

### 📊 Key Changes
- **Class Weights Support:**  
  - Added `cls_weights` option to configs for class balancing (automatic or manual).
  - Automatic calculation of class weights based on dataset distribution.
- **Weighted Sampling:**  
  - Integrated `WeightedRandomSampler` into data loading for more balanced training batches.
- **Loss Function Update:**  
  - `BCEWithLogitsLoss` now supports per-class positive weights (`pos_weight`) for better handling of imbalanced datasets.
- **Examples & Tests:**  
  - New example script (`example_class_balancing.py`) demonstrating all class balancing features.
  - Added simple and advanced tests to ensure correct class balancing behavior.
- **Backward Compatibility:**  
  - Existing configs and workflows remain unaffected if `cls_weights` is not set.

---

### 🎯 Purpose & Impact
- **Better Performance on Imbalanced Data:**  
  - Helps models learn from underrepresented classes, improving accuracy and fairness.
- **Flexible Usage:**  
  - Users can enable automatic balancing, specify custom weights, or use config files.
- **Easy Adoption:**  
  - Example scripts and thorough tests make it simple for users to understand and apply class balancing.
- **Seamless Integration:**  
  - No disruption to existing projects; new features are opt-in.

---

**In short:**  
This update makes it much easier to train YOLO models on datasets with class imbalance, leading to more reliable results—especially for rare classes! 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liam1472, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your class balancing features, please review the following checklist:

- ✅ **Define a Purpose**: Make sure your PR description clearly explains the purpose of your changes, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and follow the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is up to date with the `ultralytics/ultralytics` `main` branch. If your branch is behind, update it using the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address them accordingly.
- ✅ **Update Documentation**: Update the [documentation](https://docs.ultralytics.com/) to reflect the new or modified class balancing features.
- ✅ **Add Tests**: Ensure that you have included or updated tests to cover your changes, and confirm that all tests pass.
- ✅ **Sign the CLA**: If this is your first Ultralytics PR, please sign our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by writing "I have read the CLA Document and I sign the CLA" in a new comment.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your feature addition or bug fix. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more information, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If this is a bug fix and you haven't included a minimum reproducible example (MRE), please add one to help us verify and review the issue efficiently.

This is an automated response 🛠️. An Ultralytics engineer will review your PR and assist you further soon. Thank you for contributing to Ultralytics! 🚀🦾
